lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.cpp

```diff
@@ -54,23 +54,57 @@ Ufe::Selection ProxyShapeSceneSegmentHandler::findGatewayItems_(const Ufe::Path&
 
     // Find the MayaUSD proxyShapes
     for (const auto& stage : getAllStages()) {
-        Ufe::Path proxyShapePath = stagePath(stage);
-        // recall that findGatewayItems searches for descendents of path that are
+        const Ufe::Path proxyShapePath = stagePath(stage);
+        // recall that findGatewayItems searches for descendants of path that are
         // gateway nodes. If path itself is a gateway node it should not be included
         // in the results.
         if (proxyShapePath.startsWith(path) && proxyShapePath != path) {
             result.append(Ufe::Hierarchy::createItem(proxyShapePath));
         }
     }
 
-    // If there were Usd prims that are gateway items then we'd have an implementation
+    // TODO: If there were Usd prims that are gateway items then we'd have an implementation
     // of UsdSceneSegmentHandler that could find the gateway items and extra code
     // here to handle the case where isAGatewayType() for path is true. But right now
     // there are no gateway items in Usd, so I don't have to handle that.
 
     return result;
 }
 
+#if (UFE_PREVIEW_VERSION_NUM >= 4035)
+Ufe::Selection
+ProxyShapeSceneSegmentHandler::findGatewayItems_(const Ufe::Path& path, Ufe::Rtid nestedRtid) const
+{
+    // Handle other gateway node types that MayaUSD is not aware of.
+    if (nestedRtid != g_USDRtid) {
+        // `nestedRtid` is used as a filter. If it doesn't match the MayaUSD runtime ID, the method
+        // can return early.
+        return fMayaSceneSegmentHandler
+            ? fMayaSceneSegmentHandler->findGatewayItems_(path, nestedRtid)
+            : Ufe::Selection();
+    }
+
+    // `nestedRtid` matches the MayaUSD runtime ID. Find the MayaUSD proxyShapes.
```

Comment:
The lines 89 to 102 seems a copy & paste from lines 55 to 69. The code must be factorized in a protected (or private) method with `Ufe::Selection & result` as in & out parameter.

Reply:
I factorized the actual code to find the MayaUSD proxyShapes but left the `TODO` block as is. I feel like the `TODO` will have to be handled differently in the filtered and unfiltered method.